File: behavioral/typescript/observer_structured/interfaces.ts
export interface Observer {
  update(message: string): void;
}

export abstract class Subject {
  protected observers: Observer[] = [];
  
  attach(observer: Observer): void {
    this.observers.push(observer);
    console.log('Attached an observer');
  }
  
  detach(observer: Observer): void {
    const index = this.observers.indexOf(observer);
    if (index !== -1) {
      this.observers.splice(index, 1);
      console.log('Detached an observer');
    }
  }
  
  notify(message: string): void {
    console.log('Notifying observers...');
    for (const observer of this.observers) {
      observer.update(message);
    }
  }
}

export interface StockObserver {
  onStockPriceChange(symbol: string, price: number): void;
}

export interface WeatherObserver {
  onWeatherUpdate(temperature: number, humidity: number, pressure: number): void;
}

File: behavioral/dart/observer_structured/interfaces.dart
abstract class Observer {
  void update(String message);
}

abstract class Subject {
  final List<Observer> _observers = [];
  
  void attach(Observer observer) {
    _observers.add(observer);
    print('Attached an observer');
  }
  
  void detach(Observer observer) {
    _observers.remove(observer);
    print('Detached an observer');
  }
  
  void notify(String message) {
    print('Notifying observers...');
    for (final observer in _observers) {
      observer.update(message);
    }
  }
}

abstract class StockObserver {
  void onStockPriceChange(String symbol, double price);
}

abstract class WeatherObserver {
  void onWeatherUpdate(double temperature, double humidity, double pressure);
}
